Source/DafnyLanguageServer/Handlers/DafnyCodeActionHandler.cs

```diff
@@ -0,0 +1,157 @@
+using Microsoft.Dafny.LanguageServer.Language;
+using Microsoft.Dafny.LanguageServer.Workspace;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.Plugins;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Dafny.LanguageServer.Handlers; 
+
+public class DafnyCodeActionHandler : CodeActionHandlerBase {
+  private readonly ILogger<DafnyCodeActionHandler> logger;
+  private readonly IDocumentDatabase documents;
+
+  public DafnyCodeActionHandler(ILogger<DafnyCodeActionHandler> logger, IDocumentDatabase documents) {
+    this.logger = logger;
+    this.documents = documents;
+  }
+
+  public record QuickFixWithId(QuickFix QuickFix, int Id);
+
+  protected override CodeActionRegistrationOptions CreateRegistrationOptions(CodeActionCapability capability,
+    ClientCapabilities clientCapabilities) {
+    return new CodeActionRegistrationOptions {
+      DocumentSelector = DocumentSelector.ForLanguage("dafny"),
+      ResolveProvider = true,
+      CodeActionKinds = Container<CodeActionKind>.From(
+        CodeActionKind.QuickFix
+      ),
+      WorkDoneProgress = false
+    };
+  }
+
+
+  /// <summary>
+  /// Returns the fixes along with a unique identifier
+  /// </summary>
+  public IEnumerable<QuickFixWithId> GetFixesWithIds(QuickFixer[] fixers, DafnyDocument document, CodeActionParams request) {
+    var id = 0;
+    foreach (var fixer in fixers) {
+      var fixerInput = new VerificationQuickFixerInput(document);
+      var quickFixes = fixer.GetQuickFixes(fixerInput, request.Range);
+      var fixerCodeActions = quickFixes.Select(quickFix =>
+        new QuickFixWithId(quickFix, id++));
+      foreach (var codeAction in fixerCodeActions) {
+        yield return codeAction;
+      }
+    }
+  }
+
+  private readonly ConcurrentDictionary<string, IReadOnlyList<QuickFixWithId>> documentUriToQuickFixes = new();
+
+  public override async Task<CommandOrCodeActionContainer> Handle(CodeActionParams request, CancellationToken cancellationToken) {
+    var document = await documents.GetLastDocumentAsync(request.TextDocument);
+    if (document == null) {
+      logger.LogWarning("quick fixes requested for unloaded document {DocumentUri}", request.TextDocument.Uri);
+      return new CommandOrCodeActionContainer();
+    }
+    var quickFixers = GetQuickFixers();
+    var fixesWithId = GetFixesWithIds(quickFixers, document, request).ToArray();
+
+    var documentUri = document.Uri.ToString();
+    documentUriToQuickFixes.AddOrUpdate(documentUri,
+      _ => fixesWithId, (_, _) => fixesWithId);
+    var codeActions = fixesWithId.Select(fixWithId => {
+      CommandOrCodeAction t = new CodeAction {
```

Comment:
I'm not sure about what you mean.
The code that handles the code action is this one:

`public override Task<CodeAction> Handle(CodeAction request...`, which explicitly requires an unresolved code action. So I don't understand why I should send a command?

Reply:
I'm saying you could return a CodeAction that includes a Command, since this seems how the API is intended to be used. From the documentation: 

> A CodeAction must set either `edit` and/or a `command`. If both are supplied, the `edit` is applied first, then the `command` is executed.

Reply:
Looks like the documentation of the API is incomplete. Although what you say works well theoretically, it has the side-effect that it creates an error in VSCode.
![image](https://user-images.githubusercontent.com/3601079/184013417-27350a08-93bf-40e7-a1c9-b860ca8462b8.png)
And that's what I thought: Code actions can either give edits on the text or more general actions.
However, the API documentation you describe don't talk about the construction of code actions which is two phases, which I've seen while browing the code online
1) Creation of the code action with possible data attached
2) Resolution of the code action when it's being clicked.

In 1) there is no need to create an edit or a command if all the edits and commands are going to be computed on 2.

As a conclusion, adding a "command" would require us to change the VSCode extension to support this command, which would have to manually invoke the code action resolver... which would be useless and hard to maintain
So I'm going to revert my changes there and just use the regular data field which is there for this purpose - the data between code action creation and code action resolution.

Reply:
> As a conclusion, adding a "command" would require us to change the VSCode extension to support this command, which would have to manually invoke the code action resolver... which would be useless and hard to maintain

That indeed doesn't seem what we want.